### PR TITLE
Gather annotations for functions

### LINKED ISSUE #2895: Gather annotations for functions with impact on symbolic processing
We've recently made improvements to the `inverse` and `derivative` annotations for functions, and I think this makes it a good time to also consider the other highly related annotations for functions with impact on symbolic processing.  Currently, these are described in [18.3 _Annotations for Code Generation_](https://specification.modelica.org/master/annotations.html#annotations-for-code-generation):
- `Inline`
- `LateInline`
- `InlineAfterIndexReduction`
- `GenerateEvents`
- `smoothOrder`

As is clearly stated in the text, `smoothOrder` is closely related to `derivative`:
> A function declaration can have an annotation \fmtannotationindex{derivative} specifying the derivative function or preferably, for a function written in Modelica, use the \lstinline!smoothOrder! annotation to indicate that the tool can construct the derivative function automatically, \cref{annotations-for-code-generation}.

To me, this just shows that `smoothOrder` and `derivative` should be described next to each other.

That the inlining annotations have a close relation with `inverse` and `derivative` should probably be expressed more clearly both in text and by describing the annotations in the same chapter and section.  When to inline is seems to me as more of a tricky tradeoff between keeping the specialized `inverse` and `derivative` functions applicable and making the expressions of a function body available to general symbolic processing, than a question about how to avoid function calling overhead in the generated code (referring to the title of 18.3).

`GenerateEvents` has an intricate relation to inlining.  For example, the combination `GenerateEvents = true` and `Inline = false` is particularly challenging, showing the close relation between the two.  Because of this, but also because `GenerateEvents` has other effects on symbolic processing as well, it should be described along with annotations for inlining as well as other annotations for functions with impact on symbolic processing.

If all of the above would be moved to a common section for annotations for functions (chapter 12 or 18), this would leave only `Evaluate` and `HideResult` in 18.3, neither of which seems related to code generation.  Thus, to make a more concrete proposal, I suggest:
- Moving `smoothOrder` to 12.7 _Declaring Derivatives of Functions_
- Introducing a new section, _Function Inlining and Event Generation_ after 12.8, where we place `Inline`, `LateInline`, `InlineAfterIndexReduction`, and `GenerateEvents`.
- Moving `HideResult` to 18.4 _Annotations for Simulations_
- `Evaluate` is so fundamental that it deserves its own section right at the beginning of chapter 18: _Annotations of Utter Importance_. :)

### PR DESCRIPTION
**Pull request by henrikt-ma**
Fixes #2895.

This is still work in progress, reflected by the _Draft_ state – the PR is initiated just to show intention of having something ready for the upcoming phone meeting.

### CODE REVIEW

**Comment by henrikt-ma** [comment]
Considering that the goal is to have something ready for the upcoming phone meeting, there is no time for polishing this more before collecting feed-back.  Removing _Draft_ state.

**Comment by henrikt-ma** [comment]
Dear reviewers, as there were no comments yet I took the opportunity to improve this PR further by matching the styles of how we present the annotations in the _Functions_ chapter.  Still seeking reviews, so that we can move forward on this.

File: chapters/functions.tex [hunk 1]
```diff
@@ -1166,22 +1166,107 @@ \subsection{Casting to Record}\label{casting-to-record}
...
+For details abouts using the \lstinline!derivative! annotation, see \cref{using-the-derivative-annotation}.
+\end{semantics}
+\end{annotationdefinition}
+
+\begin{annotationdefinition}[inverse]
+\begin{synopsis}[grammar]\begin{lstlisting}
+"inverse" "(" function-inverse { "," function-inverse } ")"
+
+function-inverse :
+   IDENT "=" name "(" IDENT { "," IDENT } ")"
```
Comment (HansOlsson) [inline chapters/functions.tex:1254]: This grammar is a new change and I think it is over-specifying it, and breaking backward compatibility.
```suggestion
   IDENT "=" type-specifier function-call-args"
```
There are a number of reasons for this:

- Type-specifier allows leading "." for global names, which is clearly good.
- It allows named arguments in the call as well (good for readability).
- It allows literal values as well.

**Review by HansOlsson — CHANGES_REQUESTED** [review]
Good except for the minor change.

**Review by HansOlsson — APPROVED** [review]
Ok

**Comment by HansOlsson** [comment]
> Ok

Exept for the conflict.

File: chapters/annotations.tex [hunk 2]
```diff
@@ -371,153 +365,77 @@ \section{Annotations for Code Generation}\label{annotations-for-code-generation}
...
+\lstinline!experiment! & Simulation experiment settings & \Cref{modelica:experiment}\\
+\lstinline!HideResult! & Don't show component's simulator result & \Cref{modelica:HideResult}\\
+\lstinline!TestCase! & Information for model used as test case & \Cref{modelica:TestCase}\\
+\hline
+\end{tabular}
+\end{center}
 
-\begin{annotationdefinition}[Inline]
+\begin{annotationdefinition}[experiment]
+% henrikt-ma 2021-04: Seems strange to allow 'experiment' completely without list of options -- what would it mean?
```
Comment (eshmoylova) [inline chapters/annotations.tex:386]: Should we open a ticket about it?

**Review by eshmoylova — APPROVED** [review]
Looks good to me.

File: chapters/annotations.tex (hunk 2 above)
Comment (henrikt-ma) [reply chapters/annotations.tex:386]: Good suggestion, but it seems like this comment was lost when the PR was closed.  Opened #2985.

**Pull request merged**